src/probspecs.nim

```diff
@@ -100,13 +102,70 @@ proc findProbSpecsExercises(repo: ProbSpecsRepo, conf: Conf): seq[ProbSpecsExerc
+  ## Changes the current directory to `dir` temporarily.
+  let startDir = getCurrentDir()
   try:
-    probSpecsRepo.remove()
-    probSpecsRepo.clone()
-    probSpecsRepo.grainsWorkaround()
-    probSpecsRepo.findProbSpecsExercises(conf)
+    setCurrentDir(dir)
+    body
   finally:
-    probSpecsRepo.remove()
+    setCurrentDir(startDir)
+
+proc findProbSpecsExercises*(conf: Conf): seq[ProbSpecsExercise] =
+  const mainBranchName = "master"
+
+  if conf.probSpecsDir.isSome():
+    let probSpecsDir = conf.probSpecsDir.get()
+    logDetailed(&"Using user-provided problem-specifications dir: {probSpecsDir}")
+
+    if not dirExists(probSpecsDir):
+      showError(&"the given problem-specifications directory does not exist: '{probSpecsDir}'")
+
+    withDir probSpecsDir:
+      if execCmd("git rev-parse") != 0:
+        showError(&"the given problem-specifications directory is not a git repository: '{probSpecsDir}'")
+
+      # Exit if the working directory is not clean.
+      if execCmd("git diff-index --quiet HEAD") != 0: # Ignores untracked files.
+        echo &"\nUnstaged changes in {probSpecsDir}:"
+        discard execCmd("git status --short --untracked=no")
+        showError(&"the given problem-specifications working directory is not clean: '{probSpecsDir}'")
+
+      # Find the name of the user's remote that points to upstream (don't assume
+      # it's named 'upstream').
+      # Exit if user has no remote that points to the correct location.
+      let (remotes, errRemotes) = execCmdEx("git remote -v")
+      if errRemotes != 0:
+        showError(&"could not run `git remote -v` in the given problem-specifications directory: '{probSpecsDir}'")
+      var remoteName, remoteUrl: string
+      var foundCorrectLocation = false
+      const upstreamLocation = "github.com/exercism/problem-specifications"
+      for line in remotes.splitLines():
+        discard line.scanf("$s$w$s$+fetch)$.", remoteName, remoteUrl)
+        if remoteUrl.contains(upstreamLocation):
+          foundCorrectLocation = true
+          break
+      if not foundCorrectLocation:
+        showError(&"there is no remote that points to '{upstreamLocation}' in the given problem-specifications directory: '{probSpecsDir}'")
+
+      # For now, just exit with an error if the HEAD is not up-to-date with
+      # upstream, even if it's possible to do a fast-forward merge.
+      if execCmd(&"git fetch --quiet {remoteName} {mainBranchName}") != 0:
+        showError(&"failed to fetch `{mainBranchName}` in problem-specifications directory: '{probSpecsDir}'")
+
+      # Allow HEAD being on a non-`master` branch, as long as it's up-to-date
+      # with `upstream/master`.
+      let (revHead, _) = execCmdEx("git rev-parse HEAD")
+      let (revUpstream, _) = execCmdEx(&"git rev-parse {remoteName}/{mainBranchName}")
+      if revHead != revUpstream:
+        showError(&"the given problem-specifications directory is not up-to-date: '{probSpecsDir}'")
```

Comment:
Nice! I would suggest moving this to a separate function, as the verification code now slightly obscures the flow of the function

Reply:
Good idea. Done.

I also moved the remote-finding into its own function, but didn't do that for the rest of the validating. It think it might be better to put it into a new module if we wanted to do that.